### PR TITLE
[3/3] Applications endpoint

### PR DESCRIPTION
**Pull request by zsinx6**
This PR is a continuation of the #4
It covers for the PUT method, it is already compatible with the https://github.com/omnivector-solutions/jobbergate-cli/tree/local_fastapi

### CODE REVIEW

File: jobbergateapi2/apps/applications/routers.py [hunk 1]
```diff
@@ -148,5 +143,71 @@ async def applications_get_by_id(
...
+    )
+    async with database.transaction():
+        try:
+            await database.execute(q_update)
+
+        except INTEGRITY_CHECK_EXCEPTIONS as e:
+            raise HTTPException(status_code=422, detail=str(e))
+    application_location = (
+        f"{settings.S3_BASE_PATH}/TEST/applications/{application_id}/jobbergate.tar.gz"
+        # f"{S3_BASE_PATH}/{application.owner_id}/applications/{application.id}/jobbergate.tar.gz"
```
Comment (zsinx6) [inline jobbergateapi2/apps/applications/routers.py:199]: This is here only for testing, and the line below will be used

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 2]
```diff
@@ -245,7 +245,7 @@ async def test_get_application_all_from_user(client, user_data, application_data
 @database.transaction(force_rollback=True)
 async def test_get_application_all_from_user_empty(client, user_data, application_data):
     """
-    Return all applications from the user, even when the user don't have any
+    Return all applications from the user, even when the user don't have any.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:248]: Possible rewording
```
"""
Assert the response from GET /applications doesn't contain applications owned by other users.

This test proves that the user making the request cannot see other user's applications. We show this ​by creating applications that are owned by another user id and assert that the user making the request to list applications doesn't see any of the other user's applications in the response, len(resp.json()) == 0.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 3]
```diff
@@ -268,7 +268,7 @@ async def test_get_application_all_from_user_empty(client, user_data, applicatio
 @database.transaction(force_rollback=True)
 async def test_get_all_applications(client, user_data, application_data):
     """
-    If all=True returns all applications
+    If all=True returns all applications.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:271]: Possible rewording
```python
"""
Assert the response from GET /applications?all=True contains applications owned by other users.

This test proves that the user making the request can see applications owned by other users.
We show this ​by creating applications that are owned multiple other users and assert that
the user making the request to list applications includes the applications owned by other users.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 4]
```diff
@@ -290,3 +290,63 @@ async def test_get_all_applications(client, user_data, application_data):
...
     assert data[1]["id"] == 2
     assert data[2]["id"] == 3
+
+
+@pytest.mark.asyncio
+@mock.patch("jobbergateapi2.apps.applications.routers.boto3")
+@database.transaction(force_rollback=True)
+async def test_update_application(boto3_client_mock, client, user_data, application_data):
+    """
+    Test updating an application.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:300]: Possible rewording
```python
"""
Test that an application is updated via PUT.

This test proves that an application's values are correctly updated following a PUT request to the
/application/<id> endpoint. We show this ​by asserting that the values provided to update the
application are returned in the response made to the PUT /applciation/<id> endpoint.
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 5]
```diff
@@ -290,3 +290,63 @@ async def test_get_all_applications(client, user_data, application_data):
...
+
+    s3_client_mock.put_object.assert_called_once()
+
+
+@pytest.mark.asyncio
+@mock.patch("jobbergateapi2.apps.applications.routers.boto3")
+@database.transaction(force_rollback=True)
+async def test_update_application_wrong_user(boto3_client_mock, client, user_data, application_data):
+    """
+    Should not allow userX to update application from userY.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:334]: Possible rewording
```python
"""
Test that a userX cannot update an application owned by userY.

This test proves that userX cannot update an application owned by userY. We show this by asserting that the application id doesn't exist for the user making the request and the expected response is returned.  
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 6]
```diff
@@ -218,7 +218,7 @@ async def test_get_application_by_id_invalid(client, user_data, application_data
 @database.transaction(force_rollback=True)
 async def test_get_application_all_from_user(client, user_data, application_data):
     """
-    Return all applications from the user
+    Return all applications from the user.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:221]: """
Test that the correct application is returned for the user that created it.

This test proves that GET /application/<id> returns the correct application for the user making
the request. We show this ​by asserting that the application data in the response is what
we would expect, given the application that exists in the database, owned by the user
making the request.
"""

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 7]
```diff
@@ -205,7 +205,7 @@ async def test_get_application_by_id(client, user_data, application_data):
 @database.transaction(force_rollback=True)
 async def test_get_application_by_id_invalid(client, user_data, application_data):
     """
-    Return 404 when the application id does not exists
+    Return 404 when the application id does not exists.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:208]:
```python
"""
Test that the correct response code is returned when an application does not exist.

This test proves that GET /application/<id> returns the correct response code when the
requested application does not exist. We show this ​by asserting that the status code
returned is what we would expect given the application requested doesn't exist (404).
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 8]
```diff
@@ -181,7 +181,7 @@ async def test_delete_application_without_id(boto3_client_mock, client, user_dat
 @database.transaction(force_rollback=True)
 async def test_get_application_by_id(client, user_data, application_data):
     """
-    Return the application with the given id
+    Return the application with the given id.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:184]:
```python
"""
Test GET /applications/<id>.

This test proves that GET /applications/<id> returns the correct application, owned by
the user making the request. We show this ​by asserting that the application data
returned in the response is equal to the application data that exists in the database
for the given application id.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 9]
```diff
@@ -165,7 +165,7 @@ async def test_delete_application_not_found(boto3_client_mock, client, user_data
 @database.transaction(force_rollback=True)
 async def test_delete_application_without_id(boto3_client_mock, client, user_data, application_data):
     """
-    Don't accept DELETE in /applications/ without id
+    Don't accept DELETE in /applications/ without id.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:168]:
```python
"""
Test DELETE /applications/ without <id> returns the correct response.

This test proves that DELETE /applications returns the correct response code (405)
when an application id is not specified. We show this ​by asserting that a 405 response
code is returned for a request made without an application id.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 10]
```diff
@@ -141,7 +141,7 @@ async def test_delete_application_not_from_user(boto3_client_mock, client, user_
 @database.transaction(force_rollback=True)
 async def test_delete_application_not_found(boto3_client_mock, client, user_data, application_data):
     """
-    Do nothing if application id does not exists
+    Do nothing if application id does not exists.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:144]:
```python
"""
Test DELETE /applications/<id> the correct response code when the application doesn't exist.

This test proves that DELETE /applications/<id> returns the correct response code (404)
when the application id does not exist in the database. We show this ​by asserting that a 404 response
code is returned for a request made with an application id that doesn't exist.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 11]
```diff
@@ -117,7 +117,7 @@ async def test_delete_application(boto3_client_mock, client, user_data, applicat
 @database.transaction(force_rollback=True)
 async def test_delete_application_not_from_user(boto3_client_mock, client, user_data, application_data):
     """
-    Do nothing if current user id is not the owner of the application
+    Do nothing if current user id is not the owner of the application.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:120]: Possible rewording
```python
"""
Test that a userX cannot delete an application owned by userY.

This test proves that userX cannot delete an application owned by userY. We show this by
asserting that the application owned by userY still exists after userX tries to delete it, that
the correct response code (404) is returned and that the boto3 method is never called.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 12]
```diff
@@ -93,7 +93,7 @@ async def test_create_without_file(boto3_client_mock, application_data, client,
 @database.transaction(force_rollback=True)
 async def test_delete_application(boto3_client_mock, client, user_data, application_data):
     """
-    Test delete an application
+    Test delete an application.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py:96]: Possible rewording
```python
"""
Test DELETE /applications/<id> correctly deletes an application.

This test proves that an application is successfully deleted via a DELETE request to the
/applciations/<id> endpoint. We show this by asserting that the application
no longer exists in the database after the delete request is made, the correct status code
is returned and the correct boto3 method was called.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 13]
```diff
@@ -71,7 +71,7 @@ async def test_create_without_application_name(boto3_client_mock, application_da
 @database.transaction(force_rollback=True)
 async def test_create_without_file(boto3_client_mock, application_data, client, user_data):
     """
-    Don't create application without file
+    Don't create application without file.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py]: Possible rewording
```python
"""
Test that an application cannot be created without the required parameters.

This test proves that an application cannot be successfully created without
passing the correct, required arguments in the request. We show this by asserting that the application
does not exists in the database after the request is made, the correct status code
is returned and the correct boto3 method was not called.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 14]
```diff
@@ -48,7 +48,7 @@ async def test_create(boto3_client_mock, application_data, client, user_data):
 @database.transaction(force_rollback=True)
 async def test_create_without_application_name(boto3_client_mock, application_data, client, user_data):
     """
-    Don't create application when required value is missing
+    Don't create application when required value is missing.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py]: Possible rewording
```python
"""
Test that an application cannot be created without passing the required parameters.

This test proves that an application cannot be successfully created without
passing the required arguments in the request. We show this by asserting that the application
does not exists in the database after the request is made, the correct status code
is returned and the correct boto3 method was not called.
"""
```

File: jobbergateapi2/tests/apps/applications/test_routers.py [hunk 15]
```diff
@@ -26,7 +26,7 @@
 @database.transaction(force_rollback=True)
 async def test_create(boto3_client_mock, application_data, client, user_data):
     """
-    Test creating a application
+    Test creating a application.
```
Comment (jamesbeedy) [inline jobbergateapi2/tests/apps/applications/test_routers.py]: Possible rewording
```python
"""
Test that an application is successfully created.

This test proves that an application is successfully created. We show this by asserting
that the application exists in the database after the request is made, the correct status code
is returned and the correct boto3 method was called.
"""
```

**Review by jamesbeedy — APPROVED** [review]
Nice work! 👍

**Pull request merged**